### PR TITLE
JAMES-3534 Identity/set update should work on existing server-set Identities

### PR DESCRIPTION
**Pull request by vttranlina**
Jira https://issues.apache.org/jira/browse/JAMES-3534
resolve https://github.com/linagora/james-project/issues/4427

### CODE REVIEW

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 1]
```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
+      .map(e=> (IdentityOrigin.CUSTOM, e))
 
-  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
+    val serverSetIdentities: SFlux[(String, Identity)] = SMono.fromCallable(() => identityFactory.listIdentities(user))
+      .subscribeOn(Schedulers.elastic())
+      .flatMapMany(SFlux.fromIterable)
+      .map(e=> (IdentityOrigin.SERVER_SET, e))
+
+    SFlux.merge(Seq(customIdentities, serverSetIdentities))
+      .groupBy(_._2.id)
```
Comment (vttranlina) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:144]: The `groupBy` will block reactive. But the total number of identities for each account will not too large. 
Should it ok?

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 2]
```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
 }
 
-case class IdentityNotFoundException(id: IdentityId) extends RuntimeException(s"$id could not be found")
\ No newline at end of file
+case class IdentityNotFoundException(id: IdentityId) extends RuntimeException(s"$id could not be found")
+
+object IdentityOrigin {
+  val CUSTOM: String = "custom"
+  val SERVER_SET: String = "server_set"
+}
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:172]: Use a case object?

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala (hunk 1 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:144]: groupBy is fine.

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 3]
```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
-      .flatMapMany(SFlux.fromIterable)))
+  def list(user: Username): Publisher[Identity] = {
+    val customIdentities: SFlux[(String, Identity)] = SFlux.fromPublisher(customIdentityDao.list(user))
+      .map(e=> (IdentityOrigin.CUSTOM, e))
 
-  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
+    val serverSetIdentities: SFlux[(String, Identity)] = SMono.fromCallable(() => identityFactory.listIdentities(user))
+      .subscribeOn(Schedulers.elastic())
+      .flatMapMany(SFlux.fromIterable)
+      .map(e=> (IdentityOrigin.SERVER_SET, e))
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:141]: Missing a space between `e` and `=>`

Also we can likely find a better name than `e` and if we can't we could use `_`.

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 4]
```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
+
+    SFlux.merge(Seq(customIdentities, serverSetIdentities))
+      .groupBy(_._2.id)
+      .flatMap(_.reduce((id1, id2) => id1._1 match {
+        case IdentityOrigin.CUSTOM => id1
+        case _ => id2
+      }))
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:148]: IMO we should have an `identity.merge(otherIdentity)` method.

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 5]
```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
     }
 
-  def list(user: Username): Publisher[Identity] = SFlux.merge(Seq(
-    customIdentityDao.list(user),
-    SMono.fromCallable(() => identityFactory.listIdentities(user))
-      .subscribeOn(Schedulers.elastic())
-      .flatMapMany(SFlux.fromIterable)))
+  def list(user: Username): Publisher[Identity] = {
+    val customIdentities: SFlux[(String, Identity)] = SFlux.fromPublisher(customIdentityDao.list(user))
+      .map(e=> (IdentityOrigin.CUSTOM, e))
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:136]: Missing a space between `e` and `=>`

Also we can likely find a better name than `e` and if we can't we could use `_`.

**Review by chibenwa — COMMENTED** [review]
I think we could benefit of unit tests for `IdentityRepository`.

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala (hunk 4 above)
Comment (vttranlina) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:148]: We have no field in Identity for priority checking. IMO, we shouldn't wrap Identity or add a new field for only this logic?

Comment (chibenwa) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:148]: This code block is hard to read and would benefit from being extracted. Or simplified.

**Review by Arsnael — APPROVED** [review]

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 6]
```diff
@@ -120,15 +132,55 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
+case class IdentityNotFoundException(id: IdentityId) extends RuntimeException(s"$id could not be found")
+
+case class IdentityOriginWrapper(origin: IdentityOrigin, identity: Identity)
+
+object IdentityOriginWrapper {
+  sealed trait IdentityOrigin
+
+  case class CustomIdentityOrigin() extends IdentityOrigin
+
+  case class ServerSetIdentityOrigin() extends IdentityOrigin
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:174]: case objects

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 7]
```diff
@@ -120,15 +132,55 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
...
+    if (identityWrap1.origin.isInstanceOf[CustomIdentityOrigin]) {
+      identityWrap1
+    } else {
+      identityWrap2
+    }
+
+  def fromCustom(identity: Identity): IdentityOriginWrapper = IdentityOriginWrapper(CustomIdentityOrigin(), identity)
+
+  def fromServerSet(identity: Identity): IdentityOriginWrapper = IdentityOriginWrapper(ServerSetIdentityOrigin(), identity)
+}
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:186]: `isInstanceOf` is a Scala anti pattern.

How about pattern matching?

```
sealed trait IdentityWithOrigin {
  def identity: Identity
}
object IdentityWithOrigin {
  def merge(identity1: IdentityWithOrigin, identity2: IdentityWithOrigin): IdentityWithOrigin = 
    identity1 match {
      case CustomIdentity => identity1
      case _ => identity2
    }
}
case class CustomIdentity(identity: Identity)
case class ServerSetIdentity(identity: Identity)
```

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      .isInstanceOf(classOf[ForbiddenSendFromException])
+  }
+
+  @Test
+  def listShouldReturnCustomAndServerSetEntries(): Unit = {
+    SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
+    when(identityFactory.listIdentities(BOB)).thenReturn(List(IDENTITY1))
+
+    assertThat(SFlux.fromPublisher(testee.list(BOB)).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(IDENTITY1, SFlux(customIdentityDAO.list(BOB)).blockFirst().get)
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala:105]:
```suggestion
    val customIdentity = SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
    when(identityFactory.listIdentities(BOB)).thenReturn(List(IDENTITY1))

    assertThat(SFlux.fromPublisher(testee.list(BOB)).collectSeq().block().asJava)
      .containsExactlyInAnyOrder(IDENTITY1, customIdentity)
```

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+  @Test
+  def updateShouldSuccessWhenCustomExists(): Unit = {
+    when(identityFactory.listIdentities(BOB)).thenReturn(List())
+    SMono.fromPublisher(customIdentityDAO.save(BOB, CREATION_REQUEST)).block()
+    val identity: Identity = SFlux(customIdentityDAO.list(BOB)).blockFirst().get
+
+    assertThatCode(() => SMono.fromPublisher(testee.update(BOB, identity.id, UPDATE_REQUEST)).block())
+      .doesNotThrowAnyException()
+  }
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala:143]: What are the results of these updates?

Can I update a server set identity twice?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 10]
```diff
@@ -1191,4 +1191,134 @@ trait IdentitySetContract {
...
+           |
+           |	]
+           |}""".stripMargin)
+    .when
+      .post
+    .`then`
+      .statusCode(SC_OK)
+      .extract()
+      .jsonPath()
+      .get("methodResponses[0][1].list[0].id")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1218]: Please use a JMAP back reference to only do a single JMAP call :-)

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala [hunk 11]
```diff
@@ -135,11 +134,47 @@ class IdentityRepositoryTest {
...
+    assertThat(SFlux(testee.list(BOB)).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(Identity(id = IDENTITY1.id,
+        name = IdentityName("Bob (3)"),
+        email = BOB.asMailAddress(),
+        replyTo = Some(List(EmailAddress(Some(EmailerName("My Boss (updated)")), new MailAddress("boss-updated@domain.tld")))),
+        bcc = Some(List(EmailAddress(Some(EmailerName("My Boss 2 (updated)")), new MailAddress("boss-updated-2@domain.tld")))),
+        textSignature = TextSignature("text 2 signature"),
+        htmlSignature = HtmlSignature("html 2 signature"),
+        mayDelete = MayDeleteIdentity(true)))
+  }
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala:179]: Ok now one more test with the second update request updating just one field (like the identity name) and leaves all the other fields untouched.

(If you update all the fields getting the double update right is trivial but partial updates are trickier...)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala (hunk 10 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1218]: I don't think you can use a back ref on an update? Not sure to get the point here

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1218]: Yes maybe not doable...

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala (hunk 11 above)
Comment (quantranhong1999) [reply server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala:179]: Maybe this is what Benoit suggests?
```scala
@Test
  def updateShouldSuccessWhenSecondPartialUpdateServerSetId(): Unit = {
    when(identityFactory.listIdentities(BOB)).thenReturn(List(IDENTITY1))
    SMono.fromPublisher(testee.update(BOB, IDENTITY1.id, UPDATE_REQUEST)).block()
    val secondUpdateWithName = IdentityUpdateRequest(name = Some(IdentityNameUpdate(IdentityName("Bob (3)"))), None, None,None, None)

    assertThatCode(() => SMono.fromPublisher(testee.update(BOB, IDENTITY1.id, secondUpdateWithName)).block())
      .doesNotThrowAnyException()

    assertThat(SFlux(testee.list(BOB)).collectSeq().block().asJava)
      .containsExactlyInAnyOrder(Identity(id = IDENTITY1.id,
        name = IdentityName("Bob (3)"),
        email = BOB.asMailAddress(),
        replyTo = Some(List(EmailAddress(Some(EmailerName("My Boss (updated)")), new MailAddress("boss-updated@domain.tld")))),
        bcc = Some(List(EmailAddress(Some(EmailerName("My Boss 2 (updated)")), new MailAddress("boss-updated-2@domain.tld")))),
        textSignature = TextSignature("text 2 signature"),
        htmlSignature = HtmlSignature("html 2 signature"),
        mayDelete = MayDeleteIdentity(true)))
  }
```

**Review by quantranhong1999 — APPROVED** [review]

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala (hunk 11 above)
Comment (chibenwa) [reply server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/IdentityRepositoryTest.scala:179]: Yes that's it.

Except that we shall not require the `, None, None,None, None` : this boiler plate should be removed by the use of default values...

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala (hunk 10 above)
Comment (quantranhong1999) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1218]:
```scala
val identityId = UUID.nameUUIDFromBytes("bob@domain.tld".getBytes(StandardCharsets.UTF_8))
```
to easily get the default server set id

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1218]: Very true!

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Can you rebase and squash your commit please @vttranlina ?

**Comment by vttranlina** [comment]
rebase master & squash fixup

**Pull request merged**